# Names, Assignment, and User-Defined Functions

## Demo
Last time, we saw that Python can be used as a calculator. We might think that Python is also capable of using mathematical concepts such as `pi`.

In [1]:
pi

NameError: name 'pi' is not defined

However, when we try to call `pi`, Python gave an error!

Python actually recognize names such as `pi`, but most names are not readily available until we import them. We can import `pi` from the `math` module.

In [35]:
from math import pi

This is equivalent to telling Python to make `pi` available. Now we can use `pi` for our mathematical operations.

In [5]:
pi

3.141592653589793

In [ ]:
pi * 71 / 223

We can import other functions from `math` module, such as `sin`. `sin` is a built-in function that takes an angle.

In [34]:
from math import sin

In [3]:
sin

<function math.sin(x, /)>

In [4]:
sin(pi/2)

1.0

Above are examples of built-in functions (e.g. `sin`) and names (e.g. `pi`). Turns out we can create our own name using an **assignment statement**.

An **assignment statement** has a `name` on the left hand side and an expression on the right side. 

In [41]:
radius = 10

Python evaluates the expression and binds it to the name `radius`. Now the name `radius` is bound to the value `10`.

In [6]:
radius

10

In [7]:
2 * radius

20

We can use the name `radius` to bind other names to other values. In fact, it is possible to bind multiple names to multiple values in a single statement. Below we bind the `area` and `circ`umference of a circle to some values.

In [42]:
area, circ = pi * radius * radius, 2 * pi * radius

In [43]:
area

314.1592653589793

In [44]:
circ

62.83185307179586

Both the names `area` and `circ` are now bound to the values `314.1592653589793` and `62.83185307179586` respectively. They don't remember how those values came from. Thus, if we change the `radius` value,

In [45]:
radius = 20

and if we check again the value that was bound to `area`, turns out `area` doesn't change!

In [46]:
area

314.1592653589793

This is how assignment works. Python evaluates `area` as `pi * radius * radius`, which is 314.159. but once the values is bound, Python does not remember that `area` was defined in terms of `radius`.

Assignment statements can also be used to give names to functions. Recall we used the `max` function in the previous lecture. `max` function is built-in.

In [13]:
max(1, 2, 3)

3

if we assign the name `f` to the `max` function,

In [14]:
f = max

In [15]:
f

<function max>

`f` is now the built-in `max` function. Now we can use the name `f` as if it is the `max` function. Below we try giving arguments to the name `f`.

In [16]:
f(1, 2, 3)

3

`max` is actually a name. However, when we run Python the first time, `max` is bound to a function that takes the maximum value of is arguments. In fact, we can change the value that was bound to `max`.

In [17]:
max = 7

In [18]:
max

7

Now `max` is just the number `7`. Even though we changed the value in `max`, `f` stays unchanged.

In [19]:
f(1, 2, 3)

3

In [20]:
f

<function max>

In [21]:
f(1, 2, max)

7

We can change `max` back by assigning `f` to `max`.

In [22]:
max = f

Now `max` is back to being a function that takes the maximum of its argument.

In [23]:
max(1, 2, 3)

3

Recall in the previous lecture, we used function names for common infix operators such as `+` and `*`. These functions are available in `operator` module. If we didn't import `add` and `mul` from `operator`, `add` and `mul` would be undefined to begin with.

In [24]:
add

NameError: name 'add' is not defined

In [ ]:
mul

In [28]:
from operator import add, mul

In [ ]:
add

In [ ]:
mul

We see that we can bind names to values through `import` statements. However, this only applies to names that are built-in (e.g. `add`, `mul`, `pi`). 

Another way of binding names to value is through assignment statement, which allows
1. Assigning values to new names
2. Change the values that were already bound to a name

There is even another way: `def` statement. `def` statement allows us to create our own function. Below is an example of defining the name `square` as a function that takes in a number as the argument and returns the square of that number by multipying that number with itself,

In [29]:
def square(x):
    return mul(x, x)

In [ ]:
square

In [ ]:
square(11)

We can use a call expression as the argument for the `square` function as well,

In [30]:
square(add(3, 4))

49

And we can call nested function,

In [31]:
square(square(3))

81

We can define a function that uses another function as part of it. Below is a function `sum_squares` that takes in 2 numbers `x` and `y` and returns the sum of squared `x` and squared `y`,

In [32]:
def sum_squares(x, y):
    return square(x) + square(y)

In [33]:
sum_squares(3, 4)

25

Now recall that the `radius` and `area` are currently not in sync,

In [47]:
radius

20

In [48]:
area

314.1592653589793

Above is not the area of a circle with radius 20! It is the area of a circle with radius 10 instead. How do we keep them in sync? Use a function!

This time, we can define `area` as a function that always returns `pi * radius * radius`.

In [50]:
def area():
    return pi * radius * radius

In [51]:
area

<function __main__.area()>

In [52]:
area()

1256.6370614359173

Notice that we didn't give an arguments to the function call above. This is an example of a call expression without operands.

Now the value obtained from calling the `area` function is 1256.63, which is the area of a circle with `radius` of 20.

In [53]:
pi * radius * radius

1256.6370614359173

If the `radius` is changed, the `area` is updated!

In [54]:
radius = 10
area()

314.1592653589793

In [55]:
radius = 1
area()

3.141592653589793

A function differs from a name. A function's return expression is re-evaluated everytime it's called.

Now let's review what we have seen so far.

## Types of Expression
The types of expressions we have seen so far are:

1. `Primitive` expression. For example:
    * Number or numerical (e.g. `2`, `314`)
    * Name (e.g. `add`, `mul`, `pi`)
    * String (e.g. `'hello'`)
    
2. `Call` expression, which look like the following,
<img src = 'call.jpg' width = 500/>

    * It is also possible to use a call expression without operands (e.g. `area()`)
    * We can also do nested call expression
    * An operand can also be a call expression 
<img src = 'operand.jpg' width = 500/>

## Discussion Question 1
What is the value of the final expression in this sequence?

In [56]:
f = min
f = max
g, h = min, max
max = g

# At this point, f = max, g = min, h = max, 'max' = min
max(f(2, g(h(1, 5), 3)), 4)

#Return value should be 3

3

# Environment Diagrams
Environment diagram is a way to keep track of what's happening within the Python interpreter when it executes a program. Environments are the way in which an interpreter for a programming language keeps track of what names mean. We can think of it as a memory that keeps track of what names bound to what values. This involves drawing boxes and arrows, like what most computer scientists do.

Environment diagrams visualize the interpreter's process. They look like the following,

(Before running the cell below, make sure to install `tutormagic` extension for Jupyter Notebook by kikocorreoso. See the installation documentation [here](https://github.com/kikocorreoso/tutormagic))

In [2]:
%load_ext tutormagic

In [3]:
%%tutor --lang python3

from math import pi
tau = 2 * pi

As we can see above, we have **codes on the left side** and some **frames on the right side**. The arrows next to the codes indicate where we are in the process of execution. The frames keep track of the bindings between names and values.

### Within the codes, there are:
1. Statements and expressions
<img src = 'statement.jpg' width = 500/>

2. Arrows, which indicate evaluation order
<img src = 'arrows.jpg' width = 300/>

### Within the frames, there are `bindings between names and values`
<img src = 'names.jpg' width = 400/>
**IMPORTANT: Within a frame, a name can't be repeated**. A name can only be bound to one value. Recall when we bind the name `max` to a value, the original binding, which was a function that return the maximum of the arguments, was lost. Another example is as below,

In [61]:
x = 3
x = 5

In [62]:
x

5

`x` can only be bound to the later value, 5. `x` can't be bound to both 3 and 5 on the same time.

We can also use the online `Pythontutor` to visualize environment diagrams.

Along the course, environment diagrams are going to become more complicated but also more necessary.

## Assignment Statements
Assignment statements change the bindings between names and values in frames. Look at the environment diagram below,

In [6]:
%%tutor --lang python3

a = 1
b = 2
b, a = a + b, b

There is an **execution rule** for assignment statements:
1. Evaluate all expressions to the right of `=` from left to right
2. Bind all names to the left of `=` to the resulting values in the current frame

Below, we start at the point where Python just finished executed the line `b = 2`. At this point, the global frame shows that `a` is bound to `1`, while `b` is bound to `2`.
<img src = 'assignment.jpg' width = 800/>

Now let's apply the **execution rule** to the line that is labeled **"Next to execute"**. At the right hand side of `=`, we have:
1. `a` + `b`, which evaluates to `1` + `2` = `3`
2. `b`, which just evaluates to `2`.

Then we bind the names `b` and `a` at the left side of `=` to `3` and `2`, respectively.

<img src = 'just_executed.jpg' width = 800/>

## Discussion Question 1 Solution
Recall the following sequence of codes,

In [ ]:
f = min
f = max
g, h = min, max
max = g
max(f(2, g(h(1, 5), 3)), 4)

We can use environment diagram to visualize the execution order,

In [3]:
%%tutor --lang python3

f = min
f = max
g, h = min, max
max = g
max(f(2, g(h(1, 5), 3)), 4)

Notice that the frames at the right hand side lists all the names that were defined in the codes on the left hand side. The `max` and `min` are actually also names that are part of the global frame, but they are built-in and thus by default Python does not list them in the global frame. If Python writes all the built-in names in global frame as well, the global frame will be filled with so many built-in names.. Python write built-in names in the global frame only if they are changed from the original definition.

# Defining Functions
At first, it was explained that this course is all about abstraction. In this section, we are going to learn tools for abstraction. Recall that abstraction is the process of:
1. Take a complex thing
2. Give it a name
3. Treat it as a whole without worrying about its details

`Assignment` is a simple means of abstraction as we can **bind names to values**.

`Function definition` is a more powerful means of abstraction since we can bind names to not only values, but also **expressions** (e.g. series of statements).

We define a function with the `def` keyword.
<img src = 'function.jpg' width = 500/>

We start with giving the function a `<name>`, then we list the function's `<formal parameters>`. 
<img src = 'formal_parameters.jpg' width = 400/>

Then we write the body of the function. 
<img src = 'return.jpg' width = 270/>

The first line, and the line between the `def` and `:`, is called `function signature`. `function signature` shows how many arguments a function takes in by listing the `<formal parameters>`. The `formal parameters` are `names` that refer to the argument values that are passed in into the function.

<img src = 'signature.jpg' width = 700/>

The `function body` defines what the function does. Below is a simple example of the body of a function, a return statement. The `<return expression>` is evaluated every time the function is called.

<img src = 'body.jpg' width = 700/>

There is a procedure for evaluating `def` statements.
#### Execution procedure for `def` statements:
1. Every time there's a `def` statement, we create a new function. This function has a signature `<name>`(`<formal parameters>`)
2. Set the body of that function to be **everything indented after the first line**
3. Bind the `<name>` to that function we just created in the current frame

**IMPORTANT**: When we write the body of the function, Python **DOES NOT** execute it. When we defined the `square` function,

In [ ]:
def square(x):
    return mul(x, x)

above, when we execute the `def` statement, no multiplying happened at all.

## Calling User-Defined Functions
Functions are only useful because we can call them. In addition to execution procedures, there's also a procedure for evaluating a call expression that uses a user-defined function.

#### Procedure for calling / applying user-defined functions (version 1):
1. Add a local frame, forming a **new** environment
2. Bind the function's formal parameters to its arguments in that frame
3. Execute the body of the function in that new environment

Below is a demonstration,

In [4]:
%%tutor --lang python3

from operator import mul
def square(x):
    return mul(x, x)

square(-2)

Initially, Python binds `mul` to the built-in function and `square` to the user-defined function. When Python binds `square` to the user-defined function, it does all the 3 execution procedure steps (they are not visible in the frame). However, Python has not multiplied anything yet as Python has not executed the body of the function yet. 
<img src = 'square.jpg' width = 900/>

When we execute the line `square(-2)`, Python applies the function `square(x)` to the argument value `-2`. 
<img src = 'new_frame.jpg' width = 700/>
On the right side, Python introduced a new frame called `square`, in which Python binds the formal parameter `x` to the argument value `-2`. Python then executes the body of the function in that new frame.

<img src = 'result.jpg' width = 700/>

### Recap
Here we have a built-in function and a user-defined function. We can see the formal parameter `x` in the user-defined function.
<img src = 'built_in_user_defined.jpg' width = 300/>

We have `local frame` that was introduced in the first step of the procedure for calling / applying user-defined function,
<img src = 'local_frame.jpg' width = 300/>

The original name of the function, `square`, is used to label that local frame. 
<img src = 'original_name.jpg' width = 300/>

In the local frame, we have a binding between the formal parameter `x`, which is the name of the argument, to the argument value `-2`. 
<img src = 'formal_param_bound.jpg' width = 300/>

Note that the `Return value` is NOT a binding. It is just an annotation of the result.

Earlier, it was mentioned that a function's `signature` is important. It is important since it contains all the information needed to create a local frame.

The name `square` lets us name the local frame, while the formal parameter `x` is the name where we bind the argument value of the function.
<img src = 'squarex.jpg' width = 300/>


## Looking Up Names in Environments
We now know most of the story for user-defined function. But there's one more important piece that we have not covered: **looking up names in environments**.

Every expression that Python evaluates is evaluated in the context of an environment. 

Environments are the memory that keeps track of the correspondence between names and values. Thus, the environment knows what names are bound to what values. So far, the current environment is either:
1. The global frame alond, or
2. A local frame, followed by the global frame (Note the word "followed", indicates that there is an ordering)

#### 2 very important things:
1. An environment is a sequence of frames.

    * Recall that a frame is a binding between names and values (indicated by boxes in environmental diagrams)
    
    
2. A name, when evaluated, evaluates to the value bound to that name in the earliest frame of the current environment in which that name is found

Thus, if we want to look up the value bound to a name in an environment, we check each frame in turn. For example, to look up some `name` in the body of the `square` function,
1. Look for the `name` in the local frame first. If it's found, then we obtain its value.
2. If not found, look for it in the global frame.
    * Built-in names like `max` are in the global frame too, but we don't show them in environment diagrams

In [5]:
from operator import mul
mul(3, 4)

12

Now we will once again define the `square` function, this time with a different formal parameter,

In [6]:
def square(square):
    return mul(square, square)

In [7]:
square

<function __main__.square(square)>

Above, we see that the name `square` is bound to the function `square(square)`. If we try to `square` a number,

In [8]:
square(4)

16

It works just fine! How does this work?

In [9]:
%%tutor --lang python3

from operator import mul
def square(square):
    return mul(square, square)

square(-2)

If we look closely, the only difference between this `square(square)` function and the `square(x)` function is that the formal parameter is now called `square`. If we try to look for the name `square`, we will be able to find it in the `square` local frame without needing to look to global frame.